File: vsi/parser.py
from .ast import (
    printStatement,
    AST,
    Integer,
    Varexpr,
    BinOpexpr,
    RelOpexpr,
    Andexpr,
    Orexpr,
    Notexpr,
    AssignStatement,
    IfStatement,
    WhileStatement
)
import sys


class Parser:

    def __init__(self, tokens, pos=0):
        self.tokens = tokens
        self.pos = pos
        self.ast = AST()

    def parse(self):
        self.parse_statements(ast=self.ast)
        return self.ast

    def parse_statements(self, ast):
        while self.pos <= len(self.tokens):
            if self.found_one_of(["EOF", "else"]):
                break
            elif self.found("if"):
                self.advance()
                astnode = self.parse_if_statement()
                ast.insert_node(astnode)
            elif self.found("while"):
                self.advance()
                astnode = self.parse_while_statement()
                ast.insert_node(astnode)
            elif self.found("ID", by="type"):
                if self.token_ahead().value == ":=":
                    astnode = self.parse_var_assignment()
                else:
                    astnode = self.parse_expr()
                ast.insert_node(astnode)
            elif self.found("print"):
                self.advance()
                astnode = self.parse_print_statement()
                ast.insert_node(astnode)
            else:
                break

    def found(self, token, by="value"):
        if by == "value":
            return self.tokens[self.pos].value == token
        elif by == "type":
            return self.tokens[self.pos].type == token

    def found_one_of(self, tokens, by="value"):
        for token in tokens:
            if self.found(token, by=by):
                return True
        return False

    def expect(self, token, by="value"):
        if self.found(token, by=by):
            return True
        current_token = self.tokens[self.pos].value
        args = (token, current_token)
        err = "SyntaxError: expecting {} found {}".format(*args)
        sys.stderr.write(err + "\n")
        sys.exit(1)

    def expect_one_of(self, tokens, by="value"):
        if self.found_one_of(tokens, by=by):
            return True
        current_token = self.tokens[self.pos].value
        args = (tokens, current_token)
        err = "SyntaxError: expecting one of {} found {}".format(*args)
        sys.stderr.write(err + "\n")
        sys.exit(1)

    def optional(self, token, by="value"):
        if self.found(token, by=by):
            return True

    def advance(self, by=1):
        self.pos += by

    def token_ahead(self, by=1):
        if self.pos + by < len(self.tokens):
            return self.tokens[self.pos + by]

    def parse_if_statement(self):
        condition = self.parse_expr()
        self.expect("then")
        self.advance()
        true_stmt = AST()  # sub AST
        self.parse_statements(ast=true_stmt)
        false_stmt = None
        if self.optional("else"):
            self.advance()
            false_stmt = AST()  # sub AST
            self.parse_statements(ast=false_stmt)
        self.expect("done")
        self.advance()
        astnode = IfStatement(condition, true_stmt, false_stmt)
        return astnode

    def parse_while_statement(self):
        condition = self.parse_expr()
        self.expect("do")
        self.advance()
        body = AST()  # sub AST
        self.parse_statements(ast=body)
        self.expect("done")
        self.advance()
        astnode = WhileStatement(condition, body)
        return astnode

    def parse_var_assignment(self):
        var_name = self.tokens[self.pos].value
        self.advance()
        self.expect(":=")
        self.advance()
        var_value = self.parse_expr()
        self.expect(";")
        self.advance()
        astnode = AssignStatement(var_name, var_value)
        return astnode

    def parse_print_statement(self):
        expr = self.parse_expr()
        self.expect(";")
        self.advance()
        return printStatement(expr)

    def parse_expr(self):
        output = []
        op_stack = []
        while self.pos < len(self.tokens):
            current_token = self.tokens[self.pos]
            if current_token.specific_type == "OPERATOR":
                if (op_stack and
                        current_token.precedence <= op_stack[-1].precedence):
                    output.append(op_stack.pop())
                op_stack.append(current_token)
            elif current_token.type in ("NUMBER", "ID"):
                output.append(current_token)
            else:
                break
            self.advance()

        for op in reversed(op_stack):
            output.append(op_stack.pop())
        expr_ast = self.create_ast_from_postfix(output)
        return expr_ast

    def create_ast_from_postfix(self, postfix):
        stack = []
        for token in postfix:
            if token.specific_type == "OPERATOR":
                op = token.value
                operand_2 = stack.pop()
                operand_1 = None
                if op != "not":
                    # not operator has only one operand
                    operand_1 = stack.pop()
                astnode = self.create_ast_from_expr(op, operand_2, operand_1)
                stack.append(astnode)
            elif token.type == "NUMBER":
                token = Integer(token.value)
                stack.append(token)
            elif token.type == "ID":
                token = Varexpr(token.value)
                stack.append(token)
        final = stack[0]
        try:
            if final.type == "NUMBER":
                return Integer(final.value)
            elif final.type == "ID":
                return Varexpr(final.value)
        except AttributeError:
            return final

    def create_ast_from_expr(self, op, operand_2, operand_1):
        if op in ("+", "-", "*", "/", "%"):
            return BinOpexpr(op, operand_1, operand_2)
        elif op in ("<", ">", "<=", ">=", "!=", "=="):
            return RelOpexpr(op, operand_1, operand_2)
        elif op == "and":
            return Andexpr(operand_1, operand_2)
        elif op == "or":
            return Orexpr(operand_1, operand_2)
        elif op == "not":
            return Notexpr(operand_2)
        else:
            err = "SyntaxError: Invalid operator {}".format(op)
            sys.stderr.write(err + "\n")
            sys.exit(1)

File: Vsi.Core/Parser.cs
using System;
using System.Collections.Generic;

namespace Vsi.Core
{
    public class Parser
    {
        public List<Token> Tokens { get; }
        public int Pos { get; private set; }
        public Ast Ast { get; }

        public Parser(List<Token> tokens, int pos = 0)
        {
            Tokens = tokens;
            Pos = pos;
            Ast = new Ast();
        }

        public override string ToString()
        {
            return $"Parser([{string.Join(", ", Tokens)}], {Pos}, {Ast})";
        }

        public Ast Parse()
        {
            ParseStatements(Ast);
            return Ast;
        }

        public void ParseStatements(Ast ast)
        {
            while (Pos <= Tokens.Count)
            {
                if (FoundOneOf(new[] { "EOF", "else" }))
                {
                    break;
                }
                else if (Found("if"))
                {
                    Advance();
                    var astNode = ParseIfStatement();
                    if (astNode != null) ast.InsertNode(astNode);
                }
                else if (Found("while"))
                {
                    Advance();
                    var astNode = ParseWhileStatement();
                    if (astNode != null) ast.InsertNode(astNode);
                }
                else if (Found(TokenType.Id))
                {
                    AstNode? astNode;
                    if (TokenAhead()?.Value?.ToString() == ":=")
                    {
                        astNode = ParseVarAssignment();
                    }
                    else
                    {
                        astNode = ParseExpr();
                    }
                    if (astNode != null) ast.InsertNode(astNode);
                }
                else if (Found("print"))
                {
                    Advance();
                    var astNode = ParsePrintStatement();
                    if (astNode != null) ast.InsertNode(astNode);
                }
                else
                {
                    break;
                }
            }
        }

        public bool Found(string token)
        {
            if (Pos >= Tokens.Count) return false;
            return Tokens[Pos].Value?.ToString() == token;
        }

        public bool Found(TokenType type)
        {
            if (Pos >= Tokens.Count) return false;
            return Tokens[Pos].Type == type;
        }

        public bool Found(TokenSpecificType specificType)
        {
            if (Pos >= Tokens.Count) return false;
            return Tokens[Pos].SpecificType == specificType;
        }

        public bool FoundOneOf(string[] tokens)
        {
            foreach (var token in tokens)
            {
                if (Found(token))
                    return true;
            }
            return false;
        }

        public bool FoundOneOf(TokenType[] types)
        {
            foreach (var type in types)
            {
                if (Found(type))
                    return true;
            }
            return false;
        }

        public bool FoundOneOf(TokenSpecificType[] specificTypes)
        {
            foreach (var specificType in specificTypes)
            {
                if (Found(specificType))
                    return true;
            }
            return false;
        }

        public bool Expect(string token)
        {
            if (Found(token)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting {token} found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool Expect(TokenType type)
        {
            if (Found(type)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting {type} found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool Expect(TokenSpecificType specificType)
        {
            if (Found(specificType)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting {specificType} found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool ExpectOneOf(string[] tokens)
        {
            if (FoundOneOf(tokens)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting one of [{string.Join(", ", tokens)}] found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool ExpectOneOf(TokenType[] types)
        {
            if (FoundOneOf(types)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting one of [{string.Join(", ", types)}] found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool ExpectOneOf(TokenSpecificType[] specificTypes)
        {
            if (FoundOneOf(specificTypes)) return true;
            var currentToken = Tokens[Pos].Value;
            Console.Error.WriteLine($"SyntaxError: expecting one of [{string.Join(", ", specificTypes)}] found {currentToken}");
            Environment.Exit(1);
            return false;
        }

        public bool Optional(string token)
        {
            if (Found(token)) return true;
            return false;
        }

        public void Advance(int by = 1)
        {
            Pos += by;
        }

        public Token? TokenAhead(int by = 1)
        {
            if (Pos + by < Tokens.Count)
                return Tokens[Pos + by];
            return null;
        }

        public AstNode? ParseIfStatement()
        {
            var condition = ParseExpr();
            Expect("then");
            Advance();
            var trueStmt = new Ast();
            ParseStatements(trueStmt);
            AstNode? falseStmt = null;
            if (Optional("else"))
            {
                Advance();
                falseStmt = new Ast();
                ParseStatements((Ast)falseStmt);
            }
            Expect("done");
            Advance();
            if (condition == null) return null;
            return new IfStatement(condition, trueStmt, falseStmt);
        }

        public AstNode? ParseWhileStatement()
        {
            var condition = ParseExpr();
            Expect("do");
            Advance();
            var body = new Ast();
            ParseStatements(body);
            Expect("done");
            Advance();
            if (condition == null) return null;
            return new WhileStatement(condition, body);
        }

        public AstNode? ParseVarAssignment()
        {
            var varName = Tokens[Pos].Value?.ToString();
            Advance();
            Expect(":=");
            Advance();
            var varValue = ParseExpr();
            Expect(";");
            Advance();
            if (varName == null || varValue == null) return null;
            return new AssignStatement(varName, varValue);
        }

        public AstNode? ParsePrintStatement()
        {
            var expr = ParseExpr();
            Expect(";");
            Advance();
            if (expr == null) return null;
            return new PrintStatement(expr);
        }

        public AstNode? ParseExpr()
        {
            var output = new List<Token>();
            var opStack = new Stack<Token>();

            while (Pos < Tokens.Count)
            {
                var currentToken = Tokens[Pos];
                if (currentToken.SpecificType == TokenSpecificType.Operator)
                {
                    while (opStack.Count > 0 && currentToken.Precedence <= opStack.Peek().Precedence)
                    {
                        output.Add(opStack.Pop());
                    }
                    opStack.Push(currentToken);
                }
                else if (currentToken.Type == TokenType.Number || currentToken.Type == TokenType.Id)
                {
                    output.Add(currentToken);
                }
                else
                {
                    break;
                }
                Advance();
            }

            while (opStack.Count > 0)
            {
                output.Add(opStack.Pop());
            }

            return CreateAstFromPostfix(output);
        }

        public AstNode? CreateAstFromPostfix(List<Token> postfix)
        {
            var stack = new Stack<AstNode>();
            foreach (var token in postfix)
            {
                if (token.SpecificType == TokenSpecificType.Operator)
                {
                    var op = token.Value?.ToString();
                    if (op == null) continue;

                    var operand2 = stack.Pop();
                    AstNode? operand1 = null;
                    if (op != "not")
                    {
                        operand1 = stack.Pop();
                    }
                    var astNode = CreateAstFromExpr(op, operand2, operand1);
                    if (astNode != null)
                        stack.Push(astNode);
                }
                else if (token.SpecificType == TokenSpecificType.Int && token.Value != null)
                {
                    stack.Push(new Integer((int)token.Value));
                }
                else if (token.SpecificType == TokenSpecificType.Float && token.Value != null)
                {
                    stack.Push(new Float((double)token.Value));
                }
                else if (token.Type == TokenType.Id && token.Value != null)
                {
                    stack.Push(new VarExpr(token.Value.ToString()!));
                }
            }

            if (stack.Count == 0) return null;

            var final = stack.Peek();
            return final;
        }

        public AstNode? CreateAstFromExpr(string op, AstNode operand2, AstNode? operand1)
        {
            if (operand1 != null && (op == "+" || op == "-" || op == "*" || op == "/" || op == "%"))
                return new BinOpExpr(op, operand1, operand2);
            else if (operand1 != null && (op == "<" || op == ">" || op == "<=" || op == ">=" || op == "!=" || op == "=="))
                return new RelOpExpr(op, operand1, operand2);
            else if (operand1 != null && op == "and")
                return new AndExpr(operand1, operand2);
            else if (operand1 != null && op == "or")
                return new OrExpr(operand1, operand2);
            else if (op == "not")
                return new NotExpr(operand2);
            else
            {
                Console.Error.WriteLine($"SyntaxError: Invalid operator {op}");
                Environment.Exit(1);
                return null;
            }
        }
    }
}
